lib/utils/ensure-single-browser-tab-only.js

```diff
@@ -0,0 +1,78 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import BootWarning from '../components/boot-warning';
+
+const HEARTBEAT_DELAY = 1000;
+const clientId = uuidv4();
+console.log(clientId);
+const emptyLock = [null, -Infinity];
+const foundElectron = window.process && window.process.type;
+
+if (!foundElectron) {
+  const sesh = grabSessionLock();
+  console.log(sesh);
+  if ('lock-acquired' !== sesh) {
+    ReactDOM.render(<BootWarning />, document.getElementById('root'));
+    throw new Error('Simplenote can only be opened in one tab');
+  }
+  let keepGoing = true;
+  loop(() => {
+    if (!keepGoing) {
+      return false;
+    }
+    switch (grabSessionLock()) {
+      case 'lock-acquired':
+        return true; // keep updating the lock and look for other sessions which may have taken it
+
+      default:
+        window.location.reload();
+        return false; // stop watching - the user can proceed at their own risk
+    }
+  });
+
+  window.addEventListener('beforeunload', function() {
+    keepGoing = false;
+    const [lastClient] = JSON.parse(localStorage.getItem('session-lock'));
+    console.log(lastClient, clientId);
+    lastClient === clientId && localStorage.removeItem('session-lock');
+  });
+}
+
+function uuidv4() {
+  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
+    var r = (Math.random() * 16) | 0,
+      v = c == 'x' ? r : (r & 0x3) | 0x8;
+    return v.toString(16);
+  });
+}
+
+function loop(f, delay = HEARTBEAT_DELAY) {
+  f() && setTimeout(() => loop(f, delay), delay);
+}
+
+function grabSessionLock() {
+  const [lastClient, lastBeat] =
+    JSON.parse(localStorage.getItem('session-lock')) || emptyLock;
+  const now = Date.now();
+  // easy case - someone else clearly has the lock
+  // add some hysteresis to prevent fighting between sessions
+  if (lastClient !== clientId && now - lastBeat < HEARTBEAT_DELAY * 5) {
+    return 'lock-unavailable';
+  }
+
+  // maybe nobody clearly has the lock, let's try and set it
+  localStorage.setItem('session-lock', JSON.stringify([clientId, now]));
+
+  // hard case - localStorage is shared mutable state across sessions
+  const [thisClient, thisBeat] = JSON.parse(
+    localStorage.getItem('session-lock') || emptyLock
```

Comment:
`emptyLock` isn't JSON and `Infinity` collapses to `null` when encoded that way.

we can use a try/catch helper…

```js
const [ thisClient, thisBeat ] = parseJSON( localStorage.getItem( 'session-lock' ), emptyLock );

function parseJSON( s, fallbackValue ) {
	try {
		return JSON.parse( s );
	} catch ( e ) {
		return fallbackValue;
	}
}
```

Reply:
or go one step further…

```js
const [ thisClient, thisBeat ] = load( 'session-lock', emptyLock );
store( 'session-lock', [ clientId, now ] );

function load( name, fallbackValue ) {
	try {
		return JSON.parse( localStorage.getItem( name ) );
	} catch ( e ) {
		return fallbackValue;
	}
}

function store( name, value ) {
	localStorage.setItem( name, JSON.stringify( value ) );
}
```

Reply:
It was never intended to be inside the JSON.parse.

Reply:
I think that is overkill. But that's me.